### ISSUE TITLE
Cannot add a new rule

### ISSUE THREAD

**Issue by Graham-72** [backdrop-contrib/rules#3]
@quicksketch I am stuck with a problem here that, perhaps because of my lack of understanding of how to deal with entities, means that I would very much appreciate a suggestion to help me move forward.

When I first ported this Rules module in March 2015 I was able to create rules and wrote about this in my blog 'Linking to social media' at http://oliverweb.net/node/10. Since then there have been changes to core's entity module which I suspect have caused this problem which I am now trying to fix. Although an existing rule still works, it is however impossible to create a new one.

By comparing what is happening on a D7 site with my own BD test site, it seems the problem is at line 1106 of rules/includes/rules.core.inc which is

`
$return = entity_get_controller('rules_config')->save($this);
`

This gives a value of 1 in D7 but null in BD.

I have tried to learn more about using entities from the entity example module but this is no longer working, and I have reported this to @docwilmot in https://github.com/backdrop-contrib/examples/issues/7

Are there any other examples I might usefully learn from?

**Comment by jackaponte** [comment]
Just flagging this and @Graham-72's other open issues for the Rules module port as a high priority for the Backdrop community! I was excited last week when I received the "Backdrop CMS on the rise" email that listed the Rules module as being fully ported to Backdrop, as it's definitely a key Drupal site building tool, but we might wind up disappointing folks who come here and find things stalled.

@Graham-72, @quicksketch and others, any ideas on how folks who don't have time or skills to take on this development themselves can support these efforts?

Thanks for all the work done on this to date!

**Comment by Graham-72** [comment]
@jackaponte Thanks for flagging this. I too was worried when I received the newsletter listing Rules as fully ported. I need to get Rules working for a site that I have just converted from D7 to Backdrop so any help to resolve this would be very much appreciated.

**Comment by jackaponte** [comment]
Just rediscovered that Rules is broken, and remembering that I've recently seen our project claim that Rules is ready! Anyone else in the community have time and skills to put into this?

**Comment by Graham-72** [comment]
I can put some more time into this if someone would kindly help me overcome the hurdle regarding entities that I describe above.

**Comment by docwilmot** [comment]
Hi @Graham-72 I can try and lend some assistance, but Rules is a bit of a beast so not sure if I'll be much help, and I'm at work till late. 
Incidentally, have you checked what happens with:
`dpm(entity_get_controller('rules_config'));` and
`dpm($this)`;?
(obviously if one of those is broken the full expression wont work).

**Comment by Graham-72** [comment]
@docwilmot thanks for your offer of help. I have just done a quick test now that my test site is using Backdrop 1.5.2 and things with Rules look more promising. I shall not be able to do much more before being away for the weekend, but hope to make progress next week.

**Comment by Graham-72** [comment]
@docwilmot I have been trying to make progress on this, at the moment by trying to be able to create and delete a rule. Although a rule can be created and then deleted, each operation causes an error and halts the site. This is caused by a call on function filter_entity_insert (or filter_entity_delete) in core's filter.module, which is restricted to entities of type EntityInterface and here Rules has one of type RulesReactionRule.

Knowing almost nothing about how to handle entities, I have tried looking at 'Entity Example' in the Example module but this hits the same problem with the error report 'TypeError: Argument 1 passed to filter_entity_insert() must be an instance of EntityInterface, instance of stdClass given', and so the example module will not work either. I think this problem has existed since Backdrop 1.2. 

I have tried comparing what is happening in Backdrop with a working Rules module in D7 and the problem arises in `class EntityAPIControllerExportable extends EntityAPIController`, where for example in public function delete it gets to the line `parent::delete($ids, $transaction);`. The parent is the same for BD and D7 - `EntityAPIControllerExportable`, and the parameters are the same, $ids is the rule's machine name and $transaction is `DatabaseTransaction`.

For me, with my very limited understanding of what is going on with entities, this is a road block. I don't even understand how or why the filter module gets involved here. Any suggestions will be much appreciated. And it would be really good if Entity Example could be fixed https://github.com/backdrop-contrib/examples/issues/7

**Comment by docwilmot** [comment]
@Graham-72 this was a tricky one to track down. 
Whats happening here is that Rules is creating a new instance of `RulesReactionRule` which is not an extension of `Entity` but the *controller* is `EntityAPIControllerExportable` from Entity API (or at least the "extentity" code duplicated in the port). 

So eventually when `DefaultEntityController::save()` is called to save the new object, it invokes the `entity_insert()` hooks, passing the `RulesReactionRule` object. But when `filter_entity_insert(EntityInterface $entity)` runs, it requires and expects an `EntityInterface` object, not a `RulesReactionRule` because we've decided to do strict typing.

And things go sour.

This doesnt cause problems elsewhere because all core calls to `DefaultEntityController::save()` are passed real entities. Drupal doesn't require strict typing AFAIK. And Examples was creating a quick shortcut entity out of a StdClass object. 

Short version, comment out the line 450 `$this->invoke('insert', $entity)` and line 394 `$this->invoke('delete', $entity)` in `extentity.controller.inc`. You may need to comment out all the invoke calls actually.

This means though that no such invoke hooks can be used to detect Rules events, but I'll assume (didnt check) Rules probably has its own hooks.

**Comment by Graham-72** [comment]
@docwilmot thank you very much. I can now create and delete rules - a break through for me! Now for more testing.

**Comment by docwilmot** [comment]
Great to hear.

A point of concern with this port though: I noticed that the "extentity" files are essentially Drupal's Entity API module included nearly in totality. This isn't ideal of course; we should be using core Entity API functions. 

When I installed Rules yesterday I enabled it with my Entity View Mode PR branch checked out, which means there was a new `entity_menu()` function existing from that port, and of course I got an immediate PHP error because Rules has an `entity_menu()` function as well.

It will be a task, but "extentity" needs to go.

**Comment by Graham-72** [comment]
@docwilmot there are several entity functions that Rules uses which were not in core when I first ported this and I liaised with @quicksketch about using copies from Drupal - hence the 'extentity' files. It may be that since then more entity functions are provided in core but I don't believe they are all there. I will check this.

**Comment by docwilmot** [comment]
The main functionality that Entity API in core does not provide is the UI and Metadata Wrappers. I'd already ported the wrapper stuff (https://github.com/backdrop-contrib/entity_metadata_wrapper) and there is an active issue to port the UI functionality (https://github.com/backdrop/backdrop-issues/issues/1650). We could either then port the UI functions to a new module or bake it directly into Rules.

Then make Rules dependent on those two contrib modules (wrapper and UI).

The rest of the functions could be integrated into Rules itself rather than a separate inc file e.g add to `rules_menu()` rather than keeping as `entity_menu()`

**Comment by Graham-72** [comment]
@docwilmot thanks for this excellent suggestion that Rules should make use of entity_metadata_wrapper rather than its own extentity code. Unfortunately my first attempts to do this on a freshly loaded test site have resulted in a complete crash of Backdrop. I will persevere and try and find out what is happening.

**Comment by docwilmot** [comment]
Yep, with a beast the size of Rules, and an newish EMW port, I expect all sorts of wonderful things to happen. Sorry.

Also worked on the port of the UI functionality as well, just for this issue. Hopefully one day we can get all the moving parts working together.

**Comment by docwilmot** [comment]
@Graham-72 completed https://github.com/backdrop-contrib/entity_ui
Most of the functions in extentity files would be duplicated in the new UI and the EMW code, so I suspect it would require a softly, softly approach to fixing this. I'd suggest:
- uninstall Rules
- delete the extentity files
- install EMW and UI modules first
- then try and install Rules again and see what breaks

This will need to be a collaboration between us, since I'm sure you will find things broken in my two as well. But I think it would be for the best to get all three done right.

**Comment by Graham-72** [comment]
@docwilmot I will give it a try over the next few days. I attempted just this with EMW only and killed Backdrop. Does EMW work with anything else?

**Comment by docwilmot** [comment]
> I attempted just this with EMW only

You mean enabling only EMW without Rules or UI?

> killed Backdrop

Meaning errors? Can you say what errors those were?

**Comment by Graham-72** [comment]
In a fresh install I loaded and installed EMW. No problems.

I then loaded a doctored version of Rules, ie a version without the entity functions now provided in EMW. When I then installed this version of Rules, the site crashed - blank screen and no error messages or reports.

I'm happy to try again!

**Comment by docwilmot** [comment]
I've got Rules working after installing EMW and UI first. A few functions needed to be moved around a bit. PRs to follow later today.

**Comment by docwilmot** [comment]
Actually may not be able to get back till late, so pushed now. You can try again with the GitHub versions of EMW and UI (not the latest releases).
Delete all the extentiy files.
And you will need to add one missing function to Rules:
```
 function entity_load_multiple_by_name($entity_type, $names = FALSE, $conditions = array()) {
  $entities = entity_load($entity_type, $names, $conditions);
  $info = entity_get_info($entity_type);
  if (!isset($info['entity keys']['name'])) {
    return $entities;
  }
  return entity_key_array_by_property($entities, $info['entity keys']['name']);
}
```
Might be reasonable to make this part of EMW or rename it `rules_entity_load_multiple_by_name()` and keep it Rules-specific, but for now, for testing, try in Rules. 

So far with these changes Rules and Rules UI install fine but I dont know if it actually works 😄 
Its a start.

Edit: note `entity_has_status()` was renamed to `entity_ui_entity_has_status()` so calls to this in Rules will need to be renamed.

**Comment by klonos** [comment]
> Its a start.

👍 

> I've got Rules working after installing EMW and UI first.

```
dependencies[] = entity_metadata_wrapper:entity_metadata_wrapper
dependencies[] = entity_ui:entity_ui
```

**Comment by Graham-72** [comment]
@docwilmot 
> it requires and expects an EntityInterface object, not a RulesReactionRule because we've decided to do strict typing

I'm now getting this same problem when using EMW, when I attempt to save a new rule. 
rules.core.inc line 1106 'entity_get_controller('rules_config')->save($this);' seems to be the origin, following through to entity.controller.inc line 933 'module_invoke_all.....' in EMW.

This is beyond me I'm afraid! Interestingly (to me at least) the rule does get created in spite of this. A simlar error then happens when deleting a rule, and I then find the rule has been deleted in spite of the error.

**Comment by docwilmot** [comment]
😄 sorry @Graham-72, I expected this to happen eventually, planned to mention it, forgot.
Recall that we had to remove the invokes from the controller class (`EntityAPIController`) because the invokes were causing trouble. Now though our controller class is `EMWController` and its now calling the same invokes!

So what to do is extend `EMWController` in a new class (eg `RulesEMWController extends EMWController`) and override the `EMWController::save()` and delete the invokes in this new extended class.

That should work, I think.

**Comment by robertgarrigos** [comment]
I just found my self in this situation working with Ubercart port. Export from D7 and Import rules into BD triggers the same error. Although, as @Graham-72 says, the rules get imported anyway, as well as components.

**Comment by docwilmot** [comment]
@Graham-72 is this still not working? Thought we had this figured out.

**Comment by docwilmot** [comment]
Actually we didnt quite have it figured: the recommendation to delete invokes is wrong. The proper fix is, according to https://github.com/backdrop/backdrop-issues/issues/2558, we need to make `RulesReactionRule` a real entity.

**Comment by Graham-72** [comment]
> we need to make RulesReactionRule a real entity

@docwilmot yes, agreed, and I'm afraid I haven't yet had the time, and perhaps ability, to get back to tackling this. Rules needs quite a bit of attention.

**Comment by docwilmot** [comment]
@Graham-72 if you can clear up https://github.com/backdrop-contrib/rules/issues/8, I could push a PR for this. I think its best not to hold up Ubercart's porting.

**Comment by Graham-72** [comment]
@docwilmot thanks for the offer. I have now deleted the two files you refer to in #8.

**Comment by docwilmot** [comment]
Hopefully this is working with the new commits to EMW. I opted to make EMW check if the entity being saved extends core `Entity` before invoking core entity hooks. This seems like a reasonable temporary fix until we can get Rules entities to be real Entities.
@robertgarrigos can you check if this helps with your Ubercart work please? Pull the latest Rules and EMW to test.

**Comment by herbdool** [comment]
I created a separate issue for the EntityInterface issue @docwilmot https://github.com/backdrop-contrib/rules/issues/16 but perhaps you've found a workaround with your last comment?

**Comment by robertgarrigos** [comment]
thanks @docwilmot getting better. I can manually import ubercart rules and got only minor errors:

````
location: http://botiga.lndo.site/admin/config/workflow/rules/components
error: Theme hook "entity_status" not found.
````
this might be ubercart's problem:

````
location: http://botiga.lndo.site/admin/store/settings/payment/manage/uc_payment_method_free_order
error: Notice: Undefined index: label in RulesUICategory::getOptions() (line 1231 of /app/modules/rules/ui/ui.core.inc).
````
Also I found a typo and fixed with PR https://github.com/backdrop-contrib/rules/pull/18

I still need to check how ubercart imports rules on install

**Comment by robertgarrigos** [comment]
Tried again to import a component, after checking out my last PR branch and got only the "theme hook not found" error

**Comment by docwilmot** [comment]
Do you mean it works except for that error or that the error prevents it working?

**Comment by robertgarrigos** [comment]
It looks like it's working. Error only on logs

**Comment by Graham-72** [comment]
Closing this old issue now. Please open a new issue for any problem.
